R/otel.R

```diff
@@ -0,0 +1,159 @@
+otel_tracer_name <- "co.posit.r-package.ellmer"
+
+# Inspired by httr2:::get_tracer() / shiny:::get_tracer()
+# Using local scope avoids an environment object lookup on each call.
+default_tracer <- local({
+  tracer <- NULL
+  function() {
+    if (!is.null(tracer)) {
+      return(tracer)
+    }
+    if (testthat::is_testing()) {
+      # Don't cache the tracer in unit tests. It interferes with tracer provider
+      # injection in otelsdk::with_otel_record().
+      return(otel::get_tracer())
+    }
+    tracer <<- otel::get_tracer()
+    tracer
+  }
+})
+
+
+# Only activate the span if it is non-NULL. If activated, ensure it is
+# automatically ended when the activation scope exits. If
+# ospan_promise_domain is TRUE, also ensure that the active span is reactivated upon promise domain restoration.
+activate_and_cleanup_ospan <- function(
+  ospan,
+  activation_scope = parent.frame(),
+  ospan_promise_domain = TRUE
+) {
+  if (!is.null(ospan)) {
+    if (ospan_promise_domain) {
+      promises::local_ospan_promise_domain(activation_scope)
+    }
+    otel::local_active_span(
+      ospan,
+      end_on_exit = FALSE,
+      activation_scope = activation_scope
+    )
+    # For some reason, when using `end_on_exit = TRUE` above, an error would occur during `spn$end(status = "auto")`. When using `withr::defer()` here, it works fine.
+    # TODO: Set status?
+    withr::defer(promises::end_ospan(ospan), envir = activation_scope)
+  }
+}
+
+
+local_chat_ospan <- function(
+  provider,
+  parent_ospan = NULL,
+  local_envir = parent.frame(),
+  tracer = default_tracer()
+) {
+  chat_ospan <-
+    promises::create_ospan(
+      sprintf("chat %s", provider@model),
+      tracer = tracer,
+      options = list(
+        parent = parent_ospan,
+        kind = "CLIENT"
+      ),
+      attributes = list(
+        "gen_ai.operation.name" = "chat",
+        "gen_ai.system" = tolower(provider@name),
```

Comment:
```suggestion
        "gen_ai.provider.name" = tolower(provider@name),
```
As per the [semantic conventions](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/). This is evolving fast so it might well be a recent change.